src/service.rs

```diff
@@ -696,48 +643,26 @@ where
                     proto_id,
                     data.len()
                 );
-                for id in self.sessions.keys().cloned().collect::<Vec<SessionId>>() {
-                    self.push_session_message(id, proto_id, priority, data.clone());
-                }
-            }
-        }
-        self.distribute_to_session(cx);
-    }
-
-    /// Get the callback handle of the specified protocol
-    #[inline]
-    fn proto_handle(&mut self, session: bool, proto_id: ProtocolId) -> Option<InnerProtocolHandle> {
-        let handle = self.protocol_configs.values_mut().find_map(|proto| {
-            if proto.id() == proto_id {
-                if session {
-                    match proto.session_handle() {
-                        ProtocolHandle::Callback(handle) | ProtocolHandle::Both(handle) => Some(
-                            InnerProtocolHandle::Session((handle, proto.blocking_flag())),
-                        ),
-                        _ => None,
-                    }
-                } else {
-                    match proto.service_handle() {
-                        ProtocolHandle::Callback(handle) | ProtocolHandle::Both(handle) => Some(
-                            InnerProtocolHandle::Service((handle, proto.blocking_flag())),
-                        ),
-                        _ => None,
+                // Avoid unnecessary memory allocation
+                // Partial-borrowed problem on here
+                for control in self.sessions.values() {
+                    control.inner.incr_pending_data_size(data.len());
+                    let message_event = SessionEvent::ProtocolMessage {
+                        id: control.inner.id,
+                        proto_id,
+                        priority,
+                        data: data.clone(),
+                    };
+                    if priority.is_high() {
+                        self.high_write_buf
+                            .push_back((control.inner.id, message_event));
+                    } else {
+                        self.write_buf.push_back((control.inner.id, message_event));
```

Comment:
I don't think it's necessary to optimize here.

1. Copy `SessionId` is very cheap
2. Write the same logic multiple times, what if the logic changed later?
3. No evidence shows it can cause performance problem.